toolchain/check/handle_index.cpp

```diff
@@ -26,6 +26,68 @@ auto HandleParseNode(Context& /*context*/, Parse::IndexExprStartId /*node_id*/)
   return true;
 }
 
+// Performs bounds checking for string indexing when the index is a constant.
+static auto CheckStringIndexBounds(Context& context,
+                                   SemIR::InstId operand_inst_id,
+                                   SemIR::InstId index_inst_id,
+                                   const llvm::APInt& index_int) -> void {
+  if (index_int.isNegative()) {
+    CARBON_DIAGNOSTIC(ArrayIndexNegative, Error, "index `{0}` is negative.",
+                      TypedInt);
+    context.emitter().Emit(
+        SemIR::LocId(index_inst_id), ArrayIndexNegative,
+        {.type = context.insts().Get(index_inst_id).type_id(),
+         .value = index_int});
+    return;
+  }
+
+  auto operand_const_id = context.constant_values().Get(operand_inst_id);
+  if (!operand_const_id.is_constant()) {
+    return;
+  }
+
+  auto operand_const_inst_id =
+      context.constant_values().GetInstId(operand_const_id);
+  auto str_struct =
+      context.insts().TryGetAs<SemIR::StructValue>(operand_const_inst_id);
+  if (!str_struct) {
+    return;
+  }
+
+  auto elements = context.inst_blocks().Get(str_struct->elements_id);
+  CARBON_CHECK(elements.size() == 2, "String struct should have 2 fields.");
+
+  auto ptr_const_id = context.constant_values().Get(elements[0]);
+  auto ptr_inst_id = context.constant_values().GetInstId(ptr_const_id);
+  auto string_literal =
+      context.insts().TryGetAs<SemIR::StringLiteral>(ptr_inst_id);
+  if (!string_literal) {
+    return;
+  }
+
+  auto string_value = context.sem_ir().string_literal_values().Get(
+      string_literal->string_literal_id);
+  if (index_int.getActiveBits() > 64 ||
+      index_int.getZExtValue() >= string_value.size()) {
+    CARBON_DIAGNOSTIC(StringAtIndexOutOfBounds, Error,
+                      "string index `{0}` is past the end of the string.",
```

Comment:
How about something that says the bound in it?
```suggestion
                      "string index `{0}`is out of bounds; string has length {1}.",
```